README.md
# cookbook-sentry

Installs the [Sentry](https://docs.sentry.io/) error-logging software and all dependencies. Installation steps are taken from the official [instructions](https://docs.sentry.io/server/installation/python/).

## Supported Platforms

Only tested to work in Ubuntu 14.04

## Attributes

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['install']['user']</tt></td>
    <td>String</td>
    <td>User that runs Sentry</td>
    <td><tt>'sentry'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['install']['conf_dir']</tt></td>
    <td>String</td>
    <td>Subdirectory where Sentry conf files are placed</td>
    <td><tt>'~/.sentry'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['install']['log_dir']</tt></td>
    <td>String</td>
    <td>Subdirectory where log files are placed</td>
    <td><tt>'/var/log/sentry'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['swap']['swapfile']</tt></td>
    <td>String, Boolean False</td>
    <td>Location of swapfile to enable. This is useful for machines with not enough RAM to fully install Sentry dependencies. A `false` value disables this feature.</td>
    <td><tt>false</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['swap']['size']</tt></td>
    <td>String</td>
    <td>Size of swap file if used</td>
    <td><tt>'2G'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['deps']['packages']</tt></td>
    <td>Array</td>
    <td>OS packages to install beforehand</td>
    <td><tt>['python-dev', 'libxml2-dev', 'libxslt1-dev', 'libffi-dev', 'libjpeg-dev', 'libyaml-dev', 'libpq-dev', 'clang-3.8']</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['deps']['symlinks']</tt></td>
    <td>Hash</td>
    <td>Some symbolic links might need creating in order to prep the machine, so this is a place to set that. Each hash key-pair is of the format: 'link_name' =&gt; 'target'.</td>
    <td><tt>{ '/usr/bin/clang' =&gt; '/usr/bin/clang-3.8', '/usr/bin/clang++' =&gt; '/usr/bin/clang++-3.8' }</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['python']['modules']</tt></td>
    <td>Array</td>
    <td>Array of dependent Python modules (and optional versions) to install. This is also where the main Sentry software is installed, along with its plugins.</td>
    <td><tt>[ ['requests[security]', '2.11.1'], ['amqp', '1.4.9'], ['anyjson', '0.3.3'], ['billiard', '3.3.0.23'], ['contextlib2', '0.5.4'], ['python-utils', '2.0.0'], ['pytz', '2016.7'], ['six', '1.10.0'], ['sqlparse', '0.2.1'], ['sentry', '8.9.0'], 'sentry-github', 'sentry-gitlab', 'sentry-zabbix', 'sentry-slack' ]</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['postgresql']['db_map']</tt></td>
    <td>Array</td>
    <td>Array of hashes that designate the databases and users/roles to be created in Postgresql</td>
    <td><tt>[ { :db_name =&gt; 'sentry', :db_user =&gt; 'sentry', :db_pass =&gt; 'secretpassword' } ]</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['config']['admin_email']</tt></td>
    <td>String</td>
    <td>Email address of the first user to be created</td>
    <td><tt>'admin@example.com'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['config']['admin_pass']</tt></td>
    <td>String</td>
    <td>Password of the first user to be created</td>
    <td><tt>'secretpassword'</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['config']['cron']['cleanup_days']</tt></td>
    <td>Integer</td>
    <td>Number of days before records are considered stale whenever `sentry cleanup` is done</td>
    <td><tt>30</tt></td>
  </tr>
  <tr>
    <td><tt>['cookbook-sentry']['config']['cron']['sched']</tt></td>
    <td>String</td>
    <td>When to actually do `sentry cleanup`. Must be in crontab format.</td>
    <td><tt>'0 0 * * *'</tt></td>
  </tr>
</table>

## Usage

### PostgreSQL and Redis

Configure PostgreSQL and Redis using their corresponding cookbooks, particulary passwords, software versions, and bind addresses. See default attribute file for default example values.

### Config files template

As of Sentry v8, there are two config files: `config.yml` and `sentry.conf.py`. You can provide your own template for both these files in your wrapper cookbook, then set the value of both these attributes to your cookbook's name:

```
node['cookbook-sentry']['config']['config.yml']['cookbook']
node['cookbook-sentry']['config']['sentry.conf.py']['cookbook']
```

You can also use the following attributes to pass variables into the corresponding templates:

```
node['cookbook-sentry']['config']['config.yml']['vars']
node['cookbook-sentry']['config']['sentry.conf.py']['vars']
```

`cookbook-sentry` already provides default templates for the above config files.

### cookbook-sentry::default

Include `cookbook-sentry` in your node's `run_list` to install Sentry:

```json
{
  "run_list": [
    "recipe[cookbook-sentry::default]"
  ]
}
```

## License and Authors

Author:: Earth U (<iskitingbords @ gmail.com>)


CHANGELOG.md
## 0.1.3 - 2016-11-03
### Fixed
- In supervisor service for sentry_worker, added '%(process_num)d' in process_name when numprocs is more than 1.

## 0.1.2 - 2016-10-27
### Added
- Versions on the default Sentry plugins

## 0.1.1 - 2016-10-26
### Added
- Attribute to set concurrency for sentry worker (default is number of processors in machine).
- Attribute to set 'numprocs' for worker supervisor config (default is 1).
- Default Redis optimization options according to https://docs.sentry.io/server/performance/.

## 0.1.0 - 2016-10-26
### Added
- Initial working version of the cookbook

---
Changelog format reference: http://keepachangelog.com/en/0.3.0/


attributes/default.rb
#
# Author:: Earth U (<iskitingbords @ gmail.com>)
# Cookbook Name:: cookbook-sentry
# Attribute:: default
#
# Copyright 2016, Earth U
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

cb = 'cookbook-sentry'

default[cb]['install']['user']     = 'sentry'
default[cb]['install']['group']    = node[cb]['install']['user']
default[cb]['install']['home']     = "/home/#{node[cb]['install']['user']}"
default[cb]['install']['conf_dir'] = "#{node[cb]['install']['home']}/.sentry"
default[cb]['install']['log_dir']  = '/var/log/sentry'

# Location of swap file (e.g. '/swapfile')
default[cb]['swap']['swapfile'] = false
default[cb]['swap']['size']     = '2G'

default[cb]['deps']['packages'] = %w{
  python-dev libxml2-dev libxslt1-dev libffi-dev libjpeg-dev
  libyaml-dev libpq-dev clang-3.8
}
default[cb]['deps']['symlinks'] = {
  '/usr/bin/clang'   => '/usr/bin/clang-3.8',
  '/usr/bin/clang++' => '/usr/bin/clang++-3.8'
}

default[cb]['python']['modules'] = [
  ['requests[security]', '2.11.1'],
  ['amqp', '1.4.9'],
  ['anyjson', '0.3.3'],
  ['billiard', '3.3.0.23'],
  ['contextlib2', '0.5.4'],
  ['python-utils', '2.0.0'],
  ['pytz', '2016.7'],
  ['six', '1.10.0'],
  ['sqlparse', '0.2.1'],

  # Sentry will be installed as a pip package here:
  ['sentry', '8.9.0'],

  # Sentry plugins can be installed together as pip packages:
  ['sentry-github', '0.1.2'],
  ['sentry-gitlab', '0.2.0'],
  ['sentry-zabbix', '0.0.15'],
  ['sentry-slack', '0.5.0']
]

default[cb]['postgresql']['db_map'] = [
  {
    :db_name => 'sentry',
    :db_user => 'sentry',
    :db_pass => 'secretpassword'
  }
]

default['apt']['compile_time_update']      = true
default['build-essential']['compile_time'] = true

default['cmake']['install_method']     = 'source'
default['cmake']['source']['version']  = '3.6.2'
default['cmake']['source']['checksum'] =
  '189ae32a6ac398bb2f523ae77f70d463a6549926cde1544cd9cc7c6609f8b346'

default['nodejs']['engine'] = 'node'
default['nodejs']['repo']   = 'https://deb.nodesource.com/node_6.x'
# The latest NPM will already come with Node v6.x
#default['nodejs']['npm_packages'] = [ { 'name' => 'npm' } ]

default['redisio']['version']  = '3.2.4'
override['redisio']['servers'] = [ {
  'name'            => '_sentry',
  'address'         => '127.0.0.1',
  'port'            => '6379',
  'save'            => [],
  'maxmemory'       => '1gb',
  'maxmemorypolicy' => 'allkeys-lru'
} ]

default['postgresql']['password']['postgres']     = 'secretpassword'
default['postgresql']['config_pgtune']['db_type'] = 'web'

# The first user to be created. Usually should be a superuser.
default[cb]['config']['admin_email']     = 'admin@example.com'
default[cb]['config']['admin_pass']      = 'secretpassword'
default[cb]['config']['admin_superuser'] = true

default[cb]['config']['cron']['cleanup_days']   = 30
default[cb]['config']['cron']['sched']          = '0 0 * * *'
default[cb]['config']['cron']['mailto']         = "''"
default[cb]['config']['cron']['logrotate_opts'] = %w{
  weekly
  rotate\ 10
  missingok
  compress
  notifempty
}

default[cb]['config']['worker']['sup_numprocs'] = nil
default[cb]['config']['worker']['concurrency']  = nil

# Some constants.

default[cb]['config']['sentry_bin'] = '/usr/local/bin/sentry'
default[cb]['config']['sentry_env']['path'] =
  '/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games'
default[cb]['config']['sentry_env']['node_path'] =
  '/usr/lib/nodejs:/usr/lib/node_modules:/usr/share/javascript'
default[cb]['config']['cron']['logrotate_dir'] = '/etc/logrotate.d'

# Default config file templates

default[cb]['config']['config.yml']['cookbook'] = cb
default[cb]['config']['config.yml']['vars'] = {
  :mail_backend => 'smtp',
  :mail_host    => 'localhost',
  :mail_port    => 25,
  :mail_from    => 'info@example.com',
  :redis_host   => node['redisio']['servers'][0]['address'],
  :redis_port   => node['redisio']['servers'][0]['port'],
  :more_opts    => []
}

default[cb]['config']['sentry.conf.py']['cookbook'] = cb
default[cb]['config']['sentry.conf.py']['vars'] = {
  :db_name     => node[cb]['postgresql']['db_map'][0][:db_name],
  :db_user     => node[cb]['postgresql']['db_map'][0][:db_user],
  :db_pass     => node[cb]['postgresql']['db_map'][0][:db_pass],
  :db_host     => node['postgresql']['config']['listen_addresses'],
  :db_port     => node['postgresql']['config']['port'],
  :redis_host  => node['redisio']['servers'][0]['address'],
  :redis_port  => node['redisio']['servers'][0]['port'],
  :web_host    => '0.0.0.0',
  :web_port    => 9000,
  :web_ssl     => false,
  :web_workers => 3,
  :more_opts   => []
}
